File: algorithms/5.LongestPalindromicSubstr/python/main.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        size = len(s)
        longest_palindrome = ""
    
        def is_palindrome(candidate):
            cand_size = len(candidate)
            if cand_size == 0 or cand_size == 1: return True
            if candidate[0] == candidate[-1]:
                return is_palindrome(candidate[1:-1])
            return False

        for i in range(size+1):
            for j in range(i, size+1):
                substr = s[i:j]
                if is_palindrome(substr) and len(substr) > len(longest_palindrome):
                    longest_palindrome = substr
                
        return longest_palindrome

    

if __name__ == "__main__":
    sol = Solution()
    print(sol.longestPalindrome("aaaaa"))
    print(sol.longestPalindrome("ababab"))
    print(sol.longestPalindrome("abababa"))
    print(sol.longestPalindrome(""))
    print(sol.longestPalindrome("a"))
    print(sol.longestPalindrome("ab"))
    print(sol.longestPalindrome("abc"))
    print(sol.longestPalindrome("abba"))
    print(sol.longestPalindrome("abbaa"))
    print(sol.longestPalindrome("abbaab"))
    print(sol.longestPalindrome("cackaaaaaak"))
    print(sol.longestPalindrome("abcdbbfcba"))
    print(sol.longestPalindrome("abababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababa"))

File: algorithms/5.LongestPalindromicSubstr/cpp/main.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
private:
	static bool is_palindrome(string s) {
		auto size = s.length();
		if(size == 0 || size == 1) return true;
		if(s[0] == s[size - 1]) return is_palindrome(s.substr(1, size - 2));
		return false;
	}
public:
	string longestPalindrome(string s) {
		auto size = s.length();
		string longest_palindrome = "";

		for (int i = 0; i <= size; i++) {
			for (int j = i; j <= size; j++) {
				auto substr = s.substr(i, j-i);
				if (is_palindrome(substr) && substr.length() > longest_palindrome.length()) {
					longest_palindrome = substr;
				}
			}
		}
		
		return longest_palindrome;
	}
};

int main() {
	Solution sol;

	cout << sol.longestPalindrome("babad") << endl;
	cout << sol.longestPalindrome("cbbd") << endl;
	cout << sol.longestPalindrome("a") << endl;
	cout << sol.longestPalindrome("ac") << endl;
	cout << sol.longestPalindrome("bb") << endl;
	cout << sol.longestPalindrome("") << endl;
	cout << sol.longestPalindrome("abababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababababa") << endl;

	return 0;
}